src/table_properties.rs

```diff
@@ -0,0 +1,170 @@
+use crocksdb_ffi::{self, DBTablePropertiesCollection};
+use libc::{c_void, c_char, size_t, uint64_t};
+use std::collections::HashMap;
+use std::ffi::CStr;
+use std::slice;
+
+#[allow(dead_code)]
+#[derive(Debug)]
+pub struct TableProperties {
+    pub data_size: u64,
+    pub index_size: u64,
+    pub filter_size: u64,
+    pub raw_key_size: u64,
+    pub raw_value_size: u64,
+    pub num_data_blocks: u64,
+    pub num_entries: u64,
+    pub format_version: u64,
+    pub fixed_key_len: u64,
+    pub column_family_id: u64,
+    pub column_family_name: String,
+    pub filter_policy_name: String,
+    pub comparator_name: String,
+    pub merge_operator_name: String,
+    pub prefix_extractor_name: String,
+    pub property_collectors_names: String,
+    pub compression_name: String,
+    pub user_collected_properties: HashMap<Vec<u8>, Vec<u8>>,
+    pub readable_properties: HashMap<String, String>,
+}
+
+pub type TablePropertiesCollection = HashMap<String, TableProperties>;
+
+fn ptr_to_string(ptr: *const c_char) -> Result<String, String> {
+    unsafe {
+        match CStr::from_ptr(ptr).to_str() {
+            Ok(s) => Ok(s.to_owned()),
+            Err(e) => Err(format!("{}", e)),
+        }
+    }
+}
+
+#[repr(C)]
+struct CShallowStringsMap {
+    size: size_t,
+    keys: *const *const c_char,
+    keys_lens: *const size_t,
+    values: *const *const c_char,
+    values_lens: *const size_t,
+}
+
+impl CShallowStringsMap {
+    fn to_bytes_map(&self) -> Result<HashMap<Vec<u8>, Vec<u8>>, String> {
+        let mut res = HashMap::new();
+        unsafe {
+            let keys = slice::from_raw_parts(self.keys, self.size);
+            let keys_lens = slice::from_raw_parts(self.keys_lens, self.size);
+            let values = slice::from_raw_parts(self.values, self.size);
+            let values_lens = slice::from_raw_parts(self.values_lens, self.size);
+            for i in 0..self.size {
```

Comment:
Yes, or you can use `for ((k, klen), (v, vlen)) in keys.iter().zip(keys_lens).zip(values.iter().zip(values_lens))`.